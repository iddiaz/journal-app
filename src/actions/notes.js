import  Swal  from 'sweetalert2';
import { db } from "../firebase/firebase-config";
import { collection, addDoc, getDocs, doc, updateDoc } from "firebase/firestore";
import {types} from '../types/types';
import { loadNotes } from "../helpers/loadNotes";
import { fileUpload } from '../helpers/fileUpload';



export const startNewNote = ()=>{
  
   //redux-thunk middle. El segundo argumento obtiene toda info directamente del store   
   return async( dispatch, getState ) =>{

      const state = getState();
      console.log(state);

      const {uid} = getState().auth;
      console.log(uid);

      const newNote = {
         title: '',
         body: '',
         date: new Date().getTime()
      }

      //old
      // const doc = db.collection(`${uid}/journal/notes`).add(newNote);

      // new firebase update
      const docRef = await addDoc(collection(db,  `${ uid }`, "journal/notes"), newNote );
      console.log("Document written with ID: ", docRef.id);

      dispatch( activeNote( docRef.id, newNote ) );


   }
}

export const activeNote = ( id, note ) => ({
  
   type: types.notesActive,
   payload: {
      id,
      ...note
   }

})

export const startLoadingNotes = ( uid ) => {

   return async( dispatch ) => {
     
      const notes = await loadNotes( uid );
      dispatch( setNotes( notes ) );

   }

}


export const setNotes = ( notes ) => ({
  
   type: types.notesLoad,
   payload: notes

})


export const startSaveNote = ( note ) => {
  
   return async( dispatch, getState ) => {
     
      const {uid} = getState().auth;

      if( !note.url ){
         delete note.url;
      }
      
      const noteToFirestore = {...note };
      delete noteToFirestore.id;

      const noteRef = doc(db, `${uid}/journal/notes/${note.id}`)
      await updateDoc(noteRef,noteToFirestore);
      
      // dispatch( startLoadingNotes( uid ) );
      // dispatch( refreshNote( note.id, note ) );
      dispatch( refreshNote( note.id, noteToFirestore ) );
      
      Swal.fire('Saved', note.title, 'success' );
  

   }

}


export const refreshNote = ( id, note ) => ({
   type: types.notesUpdated,
   payload: {
      id,
      // note, || o recupero el id individual en cada objeto para el key del map.
      note:{
         id,
         ...note
      }
   }
})

export const startUploading = (file) => {
   //accion asincrona, usa redux-thunk asi. 
   return async( dispatch, getState ) => {

      //recupera del store el valor de notes.
      const { active: activeNote } = getState().notes;     

      Swal.fire({
         title: 'Uploading...',
         text: 'Please whait',
         allowOutsideClick: false,
         showConfirmButton: false,
         willOpen: ()=> {
            Swal.showLoading();
         }
      })

      const fileUrl = await fileUpload(file);

      // console.log('fileURL=>>',fileUrl);
      activeNote.url =fileUrl;

      dispatch( startSaveNote(activeNote) );

      Swal.close();


   }

}




