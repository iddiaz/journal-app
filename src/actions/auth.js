import Swal from 'sweetalert2';

import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import { types } from './../types/types';
import { finishLoading, startLoading } from './ui';
import { notesLogoutCleaning } from './notes';




export const login = (uid, displayName) => ({ 
  
      type: types.login,
      payload: {
         uid,
         displayName
      }
   
})

export const startLoginEmailPassword = ( email, password ) =>{

   return (dispatch)=>{

      const auth = getAuth();

      dispatch( startLoading() );

      signInWithEmailAndPassword( auth, email, password )
         .then( ({user}) => {
            console.log(user);
            dispatch( login( user.uid, user.displayName ) );
            console.log('finidh?')
            dispatch( finishLoading() );   
              
         }).catch( (err) => {
          
            const dataErr = JSON.stringify( err.code );   
            const errorMessage= dataErr.replace(/['"]+/g, '');        

            dispatch( finishLoading() );          
      
            Swal.fire('Error', errorMessage, 'error' );
         })
          
        
      
   }
}

export const startGoogleLogin = () => {
  
   return (dispatch) => {

      const auth = getAuth();
      signInWithPopup(auth, googleAuthProvider)
         .then( ({user}) => {
            // console.log(userCred);
            dispatch(login(user.uid, user.displayName))
         })
      // .then(({user}) =>{
      //     dispatch(login(user.uid, user.displayName))
      // });

   }


}


export const startRegisterWithEmailPasswordName = ( email, password, name ) => {

   //como es asincrona neesita retornar el callback
   //tengo acceso al dispach aqui gracias al thunk
   return ( dispatch ) => {

      const auth = getAuth();
      createUserWithEmailAndPassword( auth, email, password )
         .then( async({user}) =>{
            await updateProfile( user, {displayName: name});
            console.log(user);
            dispatch( login( user.uid, user.displayName ))

         }).catch( err =>{
            console.log(err);
            const dataErr = JSON.stringify( err.code );   
            const errorMessage= dataErr.replace(/['"]+/g, '');  
            Swal.fire('Error', errorMessage, 'error' );
         })
   
   }

}

export const startLogout = ()=>{
   return(dispatch)=>{
      const auth = getAuth();
      signOut( auth ).then( ()=>{
         console.log('logout from firestore!!');
         dispatch( logout() );
         dispatch( notesLogoutCleaning() );


      } )


   }
}

export const logout = () => ({

   type:types.logout   
   
})






