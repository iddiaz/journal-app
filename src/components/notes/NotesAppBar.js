import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from './../../actions/notes';
import { fileUpload } from './../../helpers/fileUpload';

export const NotesAppBar = () => {
   
   const dispatch = useDispatch();
   const { active: note } = useSelector( state => state.notes );

   const handleSave = ()=>{
      // console.log('NOTE??', note);
      dispatch( startSaveNote( note ) );
   }

   const handlePictureClick = () => {
      document.querySelector('#fileSelector').click();
   }

   const handleFileChange = (e)=>{
      // console.log(e.target);
      // console.log('e.target.files=>>', e.target.files);
      const file = e.target.files[0]; 

      if (file) {
         
         dispatch( startUploading(file) );
      }

   }




   return (
      <div className='notes__appbar'>
         <span>28 agosto de 2022</span>
         
         <input 
            id='fileSelector'
            type="file" 
            name='file'
            style={{display: 'none'}} 
            onChange={handleFileChange} 
         />
         

         <div>
            <button className='btn' onClick={ handlePictureClick }>Picture</button>
            <button className='btn' onClick={ handleSave }>Save</button>
         </div>
      </div>
   )
}
