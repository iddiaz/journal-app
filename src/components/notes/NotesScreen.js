import React, { useEffect, useRef } from 'react'
import { useSelector } from 'react-redux';
import { NotesAppBar } from './NotesAppBar'
import { useForm } from './../../hooks/useForm';

export const NotesScreen = () => {

   //renombra el objeto que llega del store
   const {active: note } = useSelector( state => state.notes );
   const [ formValues, handleInputChange, reset ] = useForm( note );
   const {body, title} = formValues;

   const activeId = useRef( note.id );

   useEffect(() => {

      if( note.id !== activeId.current ){
         reset(note);
         activeId.current = note.id;
      }

    
   }, [note, reset ])



   return (
      <div className='notes__main-content'>

         <NotesAppBar/>

         <div className='notes__content'>
            <input 
               type="text" 
               placeholder='Some Awesome title' 
               className='notes__title-input'
               autoComplete='off' 
               value={ title }
               onChange={handleInputChange}
            />

            <textarea 
               className='notes__textarea'
               placeholder='What happend today'
               value={ body }
               onChange={handleInputChange}
             ></textarea>

             {
               (note.url) 
               && (<div className='notes__image'>
                     <img src="https://picsum.photos/700/300" alt="iamgen" />
                  </div>)
             }


         </div>

      </div>
   )
}
