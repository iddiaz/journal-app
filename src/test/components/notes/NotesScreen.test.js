

import { React } from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';


import { NotesScreen } from './../../../components/notes/NotesScreen';
import { activeNote } from '../../../actions/notes';



jest.mock('../../../actions/notes', ()=>({
   activeNote: jest.fn(),

}));



const middlewares = [ thunk ];
const mockStore = configureStore(middlewares) 


const initState = {
   auth: {
      uid: '1',
      name:'Ivan'
   },
   ui: {
      loading: false,
      msgError: null
   },
   notes: {
      active: {
         id:'2UDDeX88Km4Ph2gU6dnS',
         date: 1642426115032,
         title:'Segunda Nota',
         body:'lorem ipsum!!!',
         url:'https://res.cloudinary.com/dl0ibqjtm/image/upload/v1642426145/k6jeebbankd7isnkp1j8.png'
      },
      notes: [] 
   }
} 

let store = mockStore( initState ); 

//Acciones asincronas hay que simular el dispatch 
store.dispatch = jest.fn();


const wrapper = mount( 

   <Provider store={ store }>
     
         <NotesScreen/> 
      
   </Provider>  
   
)


describe('Pruebas en <NotesScreen/>', ()=>{
  
   test('Debe mostrarse correactamente', ()=> {

      expect (wrapper).toMatchSnapshot();

   });

   test('Debe disparar el activeNote', () => {
      // simula cambio en la caja de texto para comprovar la activacion de activeNote
      wrapper.find('input[name="title"]').simulate('change', {
         target:{
            name: 'title',
            value: 'Hola de nuevo'
         }
      });

      expect(activeNote).toHaveBeenLastCalledWith(
         '2UDDeX88Km4Ph2gU6dnS',
         {
            body:'lorem ipsum!!!',
            title: 'Hola de nuevo',
            id:'2UDDeX88Km4Ph2gU6dnS',
            date: 1642426115032,
            url:'https://res.cloudinary.com/dl0ibqjtm/image/upload/v1642426145/k6jeebbankd7isnkp1j8.png'
         }
      );  
   });
   

});