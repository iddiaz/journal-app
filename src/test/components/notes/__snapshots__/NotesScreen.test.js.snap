// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NotesScreen/> Debe mostrarse correactamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 agosto de 2022
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some Awesome title"
      type="text"
      value="Segunda Nota"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happend today"
      value="lorem ipsum!!!"
    />
    <div
      className="notes__image"
    >
      <img
        alt="iamgen"
        src="https://res.cloudinary.com/dl0ibqjtm/image/upload/v1642426145/k6jeebbankd7isnkp1j8.png"
      />
    </div>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
